### ISSUE TITLE
Loss nan after one backward pass

### ISSUE THREAD

**Issue by kuehno** [Dao-AILab/flash-attention#311]
Hi. I am currently trying to run FlashMHA on a RTX 2070 Super with Cuda 12.2 and torch=2.1.0+cu121 installed. 

![grafik](https://github.com/HazyResearch/flash-attention/assets/73906277/8b5f4946-280c-4817-ace8-e1a49824fa0f)

Everything seems to be working fine for me except for the backward pass. The first iteration seems fine but when I try updating the model its setting all weights of the attention block to zero.

Here is a screenshot with a small example:

![grafik](https://github.com/HazyResearch/flash-attention/assets/73906277/a632f955-5d6d-4059-bce4-0394590998e0)

I am also using torch.amp.autocast as follows:

`ctx = torch.amp.autocast(device_type="cuda", dtype=torch.float16)`

Does someone know what is causing the problem?

When trying the exact same code with torch.nn.MultiheadAttention instead its working as expected.

**Comment by tmm1** [comment]
@kuehno did you find any solution? was this with flash-attn v2?

**Comment by tmm1** [comment]
i'm having a similar issue

i think it might be a known problem?

in https://github.com/Dao-AILab/flash-attention/pull/266#issuecomment-1657435190, @tridao said:

>I'm fixing a few correctness issue with the v2 backward pass for some edge cases.

**Comment by tmm1** [comment]
I had some problem with my fast-attn installation. Maybe related to cuda version (11.5) or gcc version.

I created a fresh conda environment with cuda 11.8 and recompiled flash-attn and it works now.

i realized there's a problem when i tried to run the tests and saw this:

![IMG_0329](https://github.com/Dao-AILab/flash-attention/assets/2567/6b110006-73b0-4b99-bd7d-9ec9769d3671)

**Comment by kuehno** [comment]
Hi @tmm1,

sorry for responding this late but I thought I would give it a shot a couple months later and its finally working for me. I think I had a mismatch with cuda versions and therefore a compatability issue. Thanks for also responding this quick.

![grafik](https://github.com/Dao-AILab/flash-attention/assets/73906277/ba28b63a-7e2c-430e-9125-97c5b7e42c05)
